osquery/tables/system/posix/augeas.cpp

```diff
@@ -130,50 +81,71 @@ void matchAugeasPattern(augeas* aug,
     std::string node(matches[i]);
     free(matches[i]);
 
-    Row r;
-    const char* value = nullptr;
-    int result = aug_get(aug, node.c_str(), &value);
-    if (result == 1) {
-      r["node"] = node;
-
-      if (value != nullptr) {
-        r["value"] = value;
-      }
-
-      if (!use_path) {
-        r["path"] = getSpanInfo(aug, node, context);
-      } else {
-        r["path"] = pattern;
-      }
-
-      r["label"] = getLabelInfo(aug, node, context);
-
-      results.push_back(r);
-    } else if (result < 1) {
+    const char *value = nullptr, *label = nullptr;
+    char* file = nullptr;
+    result = aug_ns_attr(aug, "matches", i, &value, &label, &file);
+    if (result == -1) {
       reportAugeasError(aug);
+      return;
+    }
+
+    std::string path;
+    if (file != nullptr) {
+      path = file;
+      path = path.substr(6);
+      // The caller is responsible for the matching memory.
+      free(file);
+    } else {
+      // The iterator is currently pointing to a folder so we extract the path from the node.
+      path = node.substr(6);
     }
+
+    Row r = {{"node", node},
+             {"value", value == nullptr ? "" : value},
+             {"path", path},
+             {"label", label == nullptr ? "" : label}};
+    results[i] = r;
   }
 
   // aug_match() allocates the matches array and expects the caller to free it.
   free(matches);
 }
 
-QueryData genAugeas(QueryContext& context) {
-  augeas* aug = aug_init(
-      nullptr, FLAGS_augeas_lenses.c_str(), AUG_NO_ERR_CLOSE | AUG_ENABLE_SPAN);
+class AugeasHandle {
+public:
+  augeas* aug;
+  bool error;
+
+  AugeasHandle() {
+    error = false;
+    this->aug = aug_init(nullptr, FLAGS_augeas_lenses.c_str(), AUG_NO_ERR_CLOSE | AUG_NO_LOAD);
```

Comment:
I'd recommend not initializing this in the constructor. You should be able to initialize on first-use.

Reply:
The whole point of the class is that it is a singleton. It's initialized statically.
This initialization doesn't actually parse the configuration files because of AUG_NO_LOAD.
I can make the initialization more complex if needed.

Reply:
Understood, but with all Singleton designs you need to be careful about when initialization happens. To protect against things like the [static order initialization fiasco](https://isocpp.org/wiki/faq/ctors#static-init-order) we should consider the most appropriate time to initialize data. My recommendation is initialize on first use, so the first time the Augeas table is called.

I recommend the Meyers pattern, and we use it in a few places like [`Registry`](https://github.com/facebook/osquery/blob/master/include/osquery/registry.h#L426). But you don't have to use this, it's just an example.